### DISCUSSION TITLE
Ongoing and future projects

### DISCUSSION THREAD

**Discussion by hermancollin** [neuropoly/axondeepseg#769] [Ideas of new features]
This is a little overview of the current projects and future R&D :seedling: 

# SAM
### New TEM model using SAM
This is an ongoing project with the segment-anything-model. The repository is located at [axondeepseg/sam_myelin_seg_TEM](https://github.com/axondeepseg/sam_myelin_seg_TEM). Despite the name of the repository, the axon segmentation pipeline is also included and now beats AxonDeepSeg's default TEM model. Myelin segmentation models are still being trained atm. This pipeline is a cascaded architecture that requires the axon centroids to predict the axon mask, and then uses this mask to generate bounding boxes used by the myelin segmentation model to predict the myelin mask. 

<img width="1075" alt="Dessin2" src="https://github.com/axondeepseg/axondeepseg/assets/83031821/0123b7aa-a443-4126-a34f-4ab1f6165f0a">

An advantage of using SAM is that its transformer-based image encoder could easily be used to train on images from multiple different contrasts. Also, it alleviates any problem related to the following issue: semantic segmentations are predicted by current UNet models and are further postprocessed into instance segmentations. The watershed algorithm used for this conversion is not ideal and often produces annoying artifacts that lowers the quality of the metrics. With SAM, we automatically get a different mask for every myelin sheath in the image, so this problem completely disappears. Overlapping masks can simply be XORed together to produce a myelin segmentation with 0 overlap, and the metrics can be directly extracted.

### Related projects
- [ ] Use this pipeline on different datasets (individually) (see https://github.com/axondeepseg/sam_myelin_seg_TEM/issues/10)
- [ ] Use this pipeline on an aggregation of all our datasets (see https://github.com/axondeepseg/sam_myelin_seg_TEM/issues/11)
- [ ] Train an object detection model to produce axon centroids automatically (makes this method fully automatic) (see https://github.com/axondeepseg/sam_myelin_seg_TEM/issues/12)
- [ ] Integrate into ADS (see https://github.com/axondeepseg/sam_myelin_seg_TEM/issues/13)

# nnUNet
### Description
[nnUNetv2](https://github.com/MIC-DKFZ/nnUNet/tree/master) is a powerful UNet-based segmentation pipeline. From my experience, it works very well, especially when data is limited. The dataset needs to be preprocessed in a certain way, but everything else is automated. nnUNet chooses the best training hyperparameters based on the data. Another important feature is model ensembling, to get the most out of a small dataset. Currently, this is one of the most promising candidate to replace ivadomed.

### Current use
An axon+myelin segmentation rabbit model (BF) and an unmyelinated axon segmentation model (TEM) are deployed. For more information and to download these models, see this repository: https://github.com/axondeepseg/nn-axondeepseg.
The inference tool to use these models was not implemented directly into ADS, but it should be.

### Related projects
- [ ] Re-train ivadomed models with nnUNetv2 and compare test metrics (see #771)
- [ ] Train nnUNetv2 on an aggregation of all our datasets (see #771)
- [ ] Integrate into ADS (see #765)

# Other projects
- [ ] Integrate the myelin cutter tool. This model predicts areas where two myelin sheaths are touching. Earlier, the problem of instance segmentation postprocessing was raised. This enhances the results produced by the watershed algorithm and thus gives better instance segs. See #768 
- [ ] Add a few QA features. See #760, #640 (old)
- [ ] Add support for unmyelinated axon morphometrics. We have an nnUNet model for unmyelinated axons and an ongoing project with Stanford for another one. Eventually, it will be necessary to also extract morphometrics from these masks. Somewhat related to #698

**Comment by jcohenadad** [comment]
Amazing! Thank you @hermancollin. Please make sure projects are listed here: https://github.com/orgs/neuropoly/projects/3

**Reply by hermancollin** [reply]
Ah! I didn't realize we were using gh projects there. We might need a new ADS category for these projects.

**Reply by jcohenadad** [reply]
> We might need a new ADS category for these projects.

yes absolutely! pls feel free to create one-- @joshuacwnewton can help if you don't know how to do this

**Reply by hermancollin** [reply]
Done.
